src/Utility/Hash.php

```diff
@@ -1146,6 +1147,179 @@ public static function mergeDiff(array $data, array $compare): array
+     * @param array $config Schema config
+     * @return array
+     */
+    protected static function validateField(
+        $value,
+        array $fieldSpec,
+        string $fieldPath,
+        array $config
+    ): array {
+        foreach ($fieldSpec['type'] as $typeKind => $type) {
+            if (is_string($typeKind)) {
+                if ($typeKind === 'array{}') {
+                    if (is_array($value)) {
+                        return static::validateShape($value, $type, $fieldPath, $config);
+                    }
+                } elseif (static::isValueOfTypeGeneric($value, $typeKind, $type)) {
+                    return [];
+                }
+            } elseif (static::isValueOfType($value, $type)) {
+                return [];
+            }
+        }
+
+        return [$fieldPath => 'Type for field does not match expected type.'];
+    }
+
+    /**
+     * Returns normalized full field specification.
+     *
+     * @param array|string $fieldData Field data
+     * @param array $config Schema config
+     * @return array
+     */
+    protected static function parseFieldSpec($fieldData, array $config): array
+    {
+        if (is_string($fieldData)) {
+            // field-spec is a single type string: ['field' => 'string']
+            return ['required' => $config['required'], 'type' => [$fieldData]];
+        }
+
+        // field-spec: ['field' => ['required' => true, 'type' => 'string']]
+        return ['required' => $fieldData['required'] ?? $config['required'], 'type' => (array)($fieldData['type'])];
+    }
+
+    /**
+     * Checks if $value maches expected $type.
+     *
+     * @param mixed $value Value to check
+     * @param string $type Type name
+     * @return bool
+     */
+    protected static function isValueOfType($value, string $type): bool
+    {
+        switch ($type) {
+            case 'array':
+                return is_array($value);
+            case 'string':
+                return is_string($value);
+            case 'float':
+                return is_float($value);
+            case 'int':
+                return is_int($value);
+            case 'bool':
+                return is_bool($value);
+            case 'null':
+                return $value === null;
+            case 'mixed':
+                return true;
+        }
+
+        return $value instanceof $type;
+    }
+
+    /**
+     * Checks if a value matches expected generic type.
+     *
+     * @param mixed $value Value to check
+     * @param string $type Type name
+     * @param array|string $elements Elements type
+     * @return bool
+     */
+    protected static function isValueOfTypeGeneric($value, string $type, $elements): bool
+    {
+        switch ($type) {
+            case 'class-string<>':
+                /** @var class-string $elements */
+                return is_a($value, $elements, true);
+            case 'list<>':
+                break;
```

Comment:
Could this be `return is_array($value) && $value == array_values($value)`. Numerically indexed arrays are about as close as we can get to 'lists'.

Reply:
I'm going to check the type of each entry for the generic one `list<>`.  I'll add a basic `list` type as well that just verifies indexed arrays.